app/utils/NPMRepositoryParser.scala

```diff
@@ -5,37 +5,55 @@ import java.io.File
 import model._
 import play.api.libs.json.{JsObject, JsString, JsValue, Json}
 
-import scala.io.Source
-
 /**
   * Parse a npm project and extract version data.
   */
 object NPMRepositoryParser extends AbstractParser {
 
   private val buildFileName = "package.json"
+  private val lockFileName = "package-lock.json"
   private val npmVersion = """[~<>=v\^]*(.*)""".r
 
   override def buildRepository(file: File, groupName: String, springBootDefaultData: SpringBootData, springBootMasterData: SpringBootData): Option[Repository] = {
     // project files
     val buildFile = getBuildFile(file)
 
-    val source: String = Source.fromFile(buildFile).getLines.mkString
+    val source = getFileAsString(buildFile)
     val jsonValues = Json.parse(source).as[JsObject].value
 
     val name = jsonValues.get("name")
       .map(_.as[JsString].value)
       .getOrElse(file.getName)
 
-    val devDependencies = extractDependencies(jsonValues.get("devDependencies"))
-      .map(p => NodeDependency(p._1, p._2))
-      .toSeq
-    val dependencies = extractDependencies(jsonValues.get("dependencies"))
+    val dependencies = (extractDependencies(jsonValues.get("devDependencies")) ++ extractDependencies(jsonValues.get("dependencies")))
       .map(p => NodeDependency(p._1, p._2))
       .toSeq
 
-    Some(Repository(name, groupName, dependencies ++ devDependencies, "NPM", Seq.empty[Plugin]))
+    val lockFile = new File(file, lockFileName)
+    val resolvedDependencies = getResolvedDependencies(lockFile)
+
+    val result = if (resolvedDependencies.isEmpty) dependencies else resolvedDependencies.filter(d => dependencies.map(_.name).contains(d.name))
+
+    Some(Repository(name, groupName, result, "NPM", Seq.empty[Plugin]))
   }
 
+  private def getResolvedDependencies(lockFile: File): Seq[NodeDependency] = {
+    if (lockFile.exists()) {
+      Json.parse(getFileAsString(lockFile))
+        .as[JsObject]
+        .value
+        .getOrElse("dependencies", JsObject.empty)
+        .asInstanceOf[JsObject]
+        .fields
+        .map(p => NodeDependency(p._1, p._2.as[JsObject].value("version").as[JsString].value))
+    } else {
+      Seq.empty
+    }
+  }
+
+  override def canProcess(repository: File): Boolean =
+    super.canProcess(repository) && !YarnLockRepositoryParser.canProcess(repository)
```

Comment:
isn't that the same as `x && !x` ? is `YarnLockRepositoryParser` missing an override of `canProcess` ?

Reply:
I'm not sure to understand your question. 
* We are in the `canProcess` of NPMRepo (override)
* We call the abstract impl (and not canProcess as it's not a reflection method)
* the abstract impl call the getBuildFile (which depends on the class)
* the getBuildFile use the private property `buildFileName`
* then we call YarnLockRepositoryParser canProcess which use the abstract impl
* the YarnLockRepositoryParser getBuildFile is used

To resume NPM can process if there is a package.json file and not a yarn-lock file 😄

Reply:
oh yeah my bad, I should have checked what the `canProcess` does 😛